### PR TITLE
[PFMENG-1371] Feat/vault ent secret sync

### PR DESCRIPTION
**Pull request by hazmei**
Special thanks to @shisheng10 and @uchinda-sph for the sharing session of their POC, issues faced and use case of secret sync.

This PR adds the following resources:
1. AWS IAM User
2. AWS IAM Group
3. AWS IAM Policy
4. AWS Access Keys
5. Time rotating resource for rotating access keys (90 days)
6. Vault Enterprise Secret Sync Destination
7. Vault Enterprise Secret Sync Association

Previously tested on web2-platform-infra webcore dev environment.

This setup provide a way to unassociate some vault secret from the secret sync as well as removing the secret sync cleanly (requires 2 step approach for removing the secret sync destination and association).

Current code only supports AWS destination for now. Will add more destinations in the future.

# Note
The secret name in aws secrets manager is not configurable. It follows the following format: `vault/<vault accessor id>/<vault secret name>`.

There's a need to have different vault_generic_endpoint resource for associating and disassociating the vault secret as the endpoint for deleting association doesn't support the `DELETE` method. See the doc [here](https://developer.hashicorp.com/vault/api-docs/system/secrets-sync#remove-association).

https://github.com/SPHTech-Platform/terraform-vault-secret-sync/blob/5dddf828e96d38ee738d393c67b6ed7333d85362/main.tf#L145-L178 

## Creating secret sync destination and association
```terraform
module "vault_secret_sync"
  source  = "app.terraform.io/sph/terraform-vault-secret-sync" # this is just an example
  version = "0.1.0-alpha" # this is just an example

  name = "webcore-dev"

  associate_secrets = {
    web2 = {
      mount       = "kvv2"
      secret_name = ["secret1","secret2"]
    }
  }
```

## Removing secret association
```terraform
module "vault_secret_sync"
  source  = "app.terraform.io/sph/terraform-vault-secret-sync" # this is just an example
  version = "0.1.0-alpha" # this is just an example

  name = "webcore-dev"

  associate_secrets = {
    web2 = {
      mount       = "kvv2"
      secret_name = ["secret1"]
    }
  }

  unassociate_secrets = {
    web2 = {
      mount       = "kvv2"
      secret_name = ["secret2"]
    }
  }
```

## Removing the secret sync destination and secret association
This requires a 2 step flow to remove the destination and association. Secret associations must be removed before the destination can be removed.

### Step 1
```terraform
module "vault_secret_sync"
  source  = "app.terraform.io/sph/terraform-vault-secret-sync" # this is just an example
  version = "0.1.0-alpha" # this is just an example

  name = "webcore-dev"

  associate_secrets = {
    web2 = {
      mount       = "kvv2"
      secret_name = ["secret1","secret2"]
    }
  }

  delete_all_secrets_associations = true
```

### Step 2
```terraform
module "vault_secret_sync"
  source  = "app.terraform.io/sph/terraform-vault-secret-sync" # this is just an example
  version = "0.1.0-alpha" # this is just an example

  name = "webcore-dev"

  associate_secrets = {
    web2 = {
      mount       = "kvv2"
      secret_name = ["secret1","secret2"]
    }
  }

  delete_all_secrets_associations = true
  delete_sync_destination         = true
```

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+      "secretsmanager:DeleteSecret",
+      "secretsmanager:RestoreSecret",
+      "secretsmanager:TagResource",
+      "secretsmanager:UntagResource",
+    ]
+    resources = [
+      "arn:aws:secretsmanager:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:secret:vault/*"
+    ]
+  }
+}
```
Comment (uchinda-sph) [inline main.tf]: can we move this block to data.tf

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  enable_mfa_enforcement = false
+  group_users            = [module.iam_user_secretsync.iam_user_name]
+
+  custom_group_policies = [
+    {
+      name   = "vault-ent-secret-sync-${random_id.this.hex}"
+      policy = data.aws_iam_policy_document.vault_ent_secrets_manager_access.json
+    }
+  ]
+}
```
Comment (uchinda-sph) [inline main.tf]: lets move this to iam.tf

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  user = module.iam_user_secretsync.iam_user_name
+
+  lifecycle {
+    create_before_destroy = true
+
+    replace_triggered_by = [
+      null_resource.rotate_access_key
+    ]
+  }
+}
```
Comment (uchinda-sph) [inline main.tf]: lets move this to iam.tf

**Review by uchinda-sph — APPROVED** [review]

**Review by shisheng10 — APPROVED** [review]
LGTM

**Review by niroz89 — COMMENTED** [review]
@hazmei add the LICENSE as well, let's publish to public registry.

**Pull request merged**